### PR TITLE
[new feature] Add HBase2 Indexer

### PR DESCRIPTION
**Pull request by mini666**
Added the ability to read and index data from HBase's tables or snapshots using Native or MapReduce.

See docs or README for more details.
docs/development/extensions-contrib/druid-hbase2-index.md

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by dkhwangbo** [comment]
@jihoonson any updates?

**Comment by jihoonson** [comment]
Sorry for the delayed review. I've been busy for the 0.17 release and now I have some time. I will try to take a look at this PR and https://github.com/apache/druid/pull/9056 sometime soon.

**Comment by jihoonson** [comment]
BTW, @mini666 @dkhwangbo have you guys had a chance to look at the new [InputSource](https://github.com/apache/druid/blob/master/core/src/main/java/org/apache/druid/data/input/InputSource.java#L55) interface? IIRC, this PR adds a new Firehose for HBase and I think it's ok because this PR was written before the new interface was added. However, the new InputSource interface does make more sense than Firehose and probably easier to implement for HBase as well. I think adding InputSource implementations would be a nice follow-up PR. :slightly_smiling_face:

**Comment by mini666** [comment]
Hello, I agree with you.
So I quickly looked at changing to InputSource.
The bottom line is that changing to an InputSource is very influential. This is the amount that almost all new classes need to be created. Of course, the actual code won't change, but it must be moved to a new method in the new class. This will also lead to testing and documentation. I think it will take a long time to finish all this work. I don't expect exactly when it will end. 
For now, in my opinion, it would be nice to include it in the main stream first and then move on.

File: extensions-contrib/druid-hbase2-indexing/src/main/java/org/apache/druid/indexer/hbase2/input/DelimiterHBaseRowKeySpec.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+public class DelimiterHBaseRowKeySpec extends HBaseRowKeySpec
+{
+
+  private final String delimiter;
+  private final byte delimiterByte;
+
+  private final List<HBaseRowKeySchema> rowkeySchemaList;
+
+  @JsonCreator
+  public DelimiterHBaseRowKeySpec(@JsonProperty("delimiter") String delimiter,
```
[the remark this answers is not in the document]
Comment (mini666) [reply extensions-contrib/druid-hbase2-indexing/src/main/java/org/apache/druid/indexer/hbase2/input/DelimiterHBaseRowKeySpec.java]: Sure!
I applied druid code style(eclipse formatter).

**Comment by jihoonson** [comment]
@mini666 are you still working on this PR?

**Pull request closed without merging**